br/pkg/restore/snap_client/pitr_collector.go

```diff
@@ -0,0 +1,472 @@
+			return err
+		}
+
+		logutil.CL(ctx).Info("Uploaded a batch of SSTs for future PiTR.",
+			zap.Duration("take", time.Since(begin)), zap.Int("file-count", fileCount))
+
+		err = c.persistExtraBackupMeta(ctx)
+		if err != nil {
+			return errors.Annotatef(err, "failed to persist backup meta when finishing batch")
+		}
+		return nil
+	}
+	return waitDone, nil
+}
+
+func (c *pitrCollector) doWithMetaLock(f func()) {
+	c.extraBackupMetaLock.Lock()
+	f()
+	c.extraBackupMetaLock.Unlock()
+}
+
+// outputPath constructs the path by a relative path for outputting.
+func (c *pitrCollector) outputPath(segs ...string) string {
+	return filepath.Join(append([]string{"v1", "ext_backups", c.name}, segs...)...)
+}
+
+func (c *pitrCollector) metaPath() string {
+	return c.outputPath("extbackupmeta")
+}
+
+func (c *pitrCollector) sstPath(name string) string {
+	return c.outputPath("sst_files", name)
+}
+
+// putSST records an SST file.
+func (c *pitrCollector) putSST(ctx context.Context, f *pb.File) error {
+	if !c.enabled {
+		return nil
+	}
+
+	begin := time.Now()
+
+	f = util.ProtoV1Clone(f)
+	out := c.sstPath(f.Name)
+
+	copier, ok := c.taskStorage.(storage.Copier)
+	if !ok {
+		return errors.Annotatef(berrors.ErrInvalidArgument, "storage %T does not support copying", c.taskStorage)
+	}
+	spec := storage.CopySpec{
+		From: f.GetName(),
+		To:   out,
+	}
+
+	copyStart := time.Now()
+	if err := copier.CopyFrom(ctx, c.restoreStorage, spec); err != nil {
+		return errors.Annotatef(err, "failed to copy sst file %s to %s, "+
+			"you may check whether permissions are granted in both %s and %s, "+
+			"and the two storages are provided by the same cloud vendor",
+			spec.From, spec.To, c.restoreStorage.URI(), c.taskStorage.URI())
+	}
+	log.Info("Copy SST to log backup storage success.", zap.String("file", f.Name), zap.Stringer("takes", time.Since(copyStart)))
+
+	f.Name = out
+	c.doWithMetaLock(func() { c.extraBackupMeta.msg.Files = append(c.extraBackupMeta.msg.Files, f) })
+
+	metrics.RestoreUploadSSTForPiTRSeconds.Observe(time.Since(begin).Seconds())
+	return nil
+}
+
+// putRewriteRule records a rewrite rule.
+func (c *pitrCollector) putRewriteRule(_ context.Context, oldID int64, newID int64) error {
+	if !c.enabled {
+		return nil
+	}
+	var err error
+	c.doWithMetaLock(func() {
+		if oldVal, ok := c.extraBackupMeta.rewrites[oldID]; ok && oldVal != newID {
+			err = errors.Annotatef(
+				berrors.ErrInvalidArgument,
+				"pitr coll rewrite rule conflict: we had %v -> %v, but you want rewrite to %v",
+				oldID,
+				oldVal,
+				newID,
+			)
+			return
+		}
+		c.extraBackupMeta.rewrites[oldID] = newID
+	})
+	return err
+}
+
+// doPersistExtraBackupMeta writes the current content of extra backup meta to the external storage.
+// This isn't goroutine-safe. Please don't call it concurrently.
+func (c *pitrCollector) doPersistExtraBackupMeta(ctx context.Context) (err error) {
```

Comment:
why not check enable() here?

Reply:
In fact only `onBatch` and `close` might be called with `enabled == false`... Other places are optional defensive code.